### ISSUE TITLE
Cleanup The Identity Param

### ISSUE THREAD

**Issue by shahzadlone** [sourcenetwork/defradb#2439]
Andy's suggestion in https://github.com/sourcenetwork/defradb/pull/2338#discussion_r1534171165:

> thought: This new param does dirty the interfaces a little bit IMO I'm fine with this being out of scope for now, but in a follow up I would like to see something like what we did for transactions - with a WithIndentity() DB func on the DB interface. Maybe add a ticket if not doing in this PR.

John's Suggestion:

> For this, you currently pass the "identity" as an argument, but in reality, it should be passed in as a Context value
eg: you have col.CreateMany(cmd.Context(), identity, docs), where you are passing in the identity value, as a param. But to make it cleaner, you should create a new context with the identity value injected into it

**Comment by jsimnz** [comment]
By request, here is a example of how the identity can be propagated via the context, and how that would change the call sites, and clients currently implemented

https://go.dev/play/p/izSWN4fMoLl

Posting here too ;)
```go
/** The following is defined in the `acp` package *//
// custom type to ensure access is gated by the `acp` package
// This is best practise for context values, see here: https://go.dev/blog/contex
type contextKey struct{} // this can be a struct or a string, doesn't matter.
const identityCtxKey contextKey = 0

func WithIdentity(ctx context.Context, identity string) context.Context {
	return context.WithValue(identityCtxKey, identity)
}

// note this can be made into an Option instead of a `(string, bool)` return
func IdentityFromContext(ctx context.Context) (string, bool) {
	identity, ok := ctx.Value(identityCtxKey).(string)
	return identity, ok
}
/** The above is defined in the `acp` package *//


/** The following are examples inside the `db`,`fetcher`, and `planner` packages where ACP calls are made *//

// collection.create
// Note: all the other `identity` propogation that happens in this func currently (ie: `c.exists`, and `c.save` no longer
// need to be here as they also propgate the identity through the context to their respective call sites)
func (c *collection) create(ctx context.Context, txn datastore.Txn, doc *client.Document) error {
	identity, ok := acp.IdentityFromContext(ctx)
	...
	return c.registerDocWithACP(ctx, identity, doc.ID().String())
}

// collection.exists
func (c *Collection) exists(ctx context.Context, txn datastore.Txn, primaryKey core.PrimaryDataStoreKey) (exists bool, isDeleted bool, err error) {
	identity, ok := acp.IdentityFromContext(ctx)
	canRead, err := c.checkAccessOfDocument(
		ctx,
		identity,
		acp.ReadPermission,
		primaryKey.DocID,
	)
	...
}

// fetcher.init
func (df *DocumentFetcher) init(
	ctx context.Context, // this arg is added beacuse the `docFetcher.init` actually doesn't have it atm but `docFetcher.Init` does.
	acp immutable.Option[acp.ACPModule],
	col client.Collection,
	fields []client.FieldDefinition,
	filter *mapper.Filter,
	docMapper *core.DocumentMapping,
	reverse bool,
) error {
	identity, ok := acp.IdentityFromContext(ctx)
	df.identity = identity
	...
}

// planner.New
func New(
	ctx context.Context,
	acp immutable.Option[acp.ACPModule],
	db client.Store,
	txn datastore.Txn,
) *Planner {
	// no identity is needed here, because its embedded inside the `ctx` object, which is used
	// within `createNode`, `deleteNode`, `scanNode`, and `updateNode`. All of which
	// just pass the `identity` value they were using into the respective `collection.X` calls
	return &Planner{
		txn:      txn,
		acp:      acp,
		db:       db,
		ctx:      ctx,
	}
}

/** The aboveare examples inside the `db`,`fetcher`, and `planner` packages where ACP calls are made *//

// ** IMPORTANT for proper context propagation over HTTP, the following is implemnted inside the `http` package **//
// http client
func (c *httpClient) setDefaultHeaders(req *http.Request) {
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/json")

	if c.txValue != "" {
		req.Header.Set(TX_HEADER_NAME, c.txValue)
	}
	
	identity, ok := acp.IdentityFromContext(req.Context()
	if ok {
		// NOTE: This stuff will change with proper identity signature work, so this is temporary but not
		// far off the end result
		req.Header.Set(IDENTITY_HEADER_NAME, identity
	}
}

// http server middleware (new middleware)
func IdentityMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(rw http.ResonseWriter, req *http.Request) {
		identityValue := req.Header.Get(IDENTITY_HEADER_NAME)
		if identityValue == "" {
			next.ServeHTTP(rw, req)
			return
		}
		
		ctx := acp.WithIdentity(req.Context(), identityValue)
		next.ServeHTTP(rw, req.WithContext(ctx))
	})
}

// ** IMPORTANT for proper context propagation over HTTP, the above is implemnted inside the `http` package **//


// exmaple usage
func main() {
	rootstore := memory.New()
	db := NewDB(rootstore)
	
	ctx := acp.WithIdentity(context.Background(), "sourceABC123")
	collection := db.GetCollectionByName(ctx, "users")      // here is a good example of why context propagation is a good idea
								// when in the future we add AdminACP, even the `db.GetCollectionX` call
								// could be protected by ACP, which with the current design, would require
								// the `identity` argument to exist on these functions to, etc.. etc..
	doc, err := collection.Get(ctx, DocID("bae-123")
}

// example CLI - NOTE: See http client/middleware for how the identity is propagated from the CLI to the DB
func MakeCollectionGetCommand() *cobra.Command {
	var identityValue string
	cmd := &cobra.Command{
		...
		RunE: func(cmd *cobra.Command, args []string) error {
			identity := acpIdentity.NewIdentity(identityValue)
			ctx := acp.WithIdentity(cmd.Context(), identity)
			col, ok := tryGetContextCollection(cmd)
			doc, err := col.Get(ctx, docID)
			...
		}
	)
	cmd.Flags().StringVarP(
		&identityValue,
		...
	)
	return cmd
}
```

**Comment by jsimnz** [comment]
Pros: ***MUCH*** cleaner propagation
Cons: Somewhat hides identity. *But* This is prob a good thing overall, since the end game of identity/acp is to use it *everywhere* basically, moreover it is specifically request scoped information, which is what context values are designed for.

**Comment by nasdf** [comment]
We've come to consensus on the approach of using context to propagate identity. All public db methods will have the identity param removed, expect for the `AddPolicy` func.
